kms/uri/uri.go

```diff
@@ -0,0 +1,86 @@
+package uri
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// URI implements a parser for a URI format based on the the PKCS #11 URI Scheme
+// defined in https://tools.ietf.org/html/rfc7512
+//
+// These URIs will be used to define the key names in a KMS.
+type URI struct {
+	*url.URL
+	Values url.Values
+}
+
+// New creates a new URI from a scheme and key-value pairs.
+func New(scheme string, values url.Values) *URI {
+	return &URI{
+		URL: &url.URL{
+			Scheme: scheme,
+			Opaque: strings.ReplaceAll(values.Encode(), "&", ";"),
+		},
+		Values: values,
+	}
+}
+
+// NewFile creates an uri for a file.
+func NewFile(path string) *URI {
+	return &URI{
+		URL: &url.URL{
+			Scheme: "file",
+			Path:   path,
+		},
+	}
+}
+
+// HasScheme returns true if the given uri has the given scheme, false otherwise.
+func HasScheme(scheme, rawuri string) bool {
+	u, err := url.Parse(rawuri)
+	if err != nil {
+		return false
+	}
+	return strings.EqualFold(u.Scheme, scheme)
+}
+
+// Parse returns the URI for the given string or an error.
+func Parse(rawuri string) (*URI, error) {
+	u, err := url.Parse(rawuri)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error parsing %s", rawuri)
+	}
+	if u.Scheme == "" {
+		return nil, errors.Errorf("error parsing %s: scheme is missing", rawuri)
+	}
+	v, err := url.ParseQuery(u.Opaque)
```

Comment:
I don't think this will work if there's more than one `pk11-pattr` in the thing you're trying to parse. ParseQuery doesn't look for `;`, only `?`.

For example, a PKCS11 uri can look like:
```
pkcs11:foo=bar;baz=qux?nob=bim&xur=wyz
```

Reply:
In that case `u.Opaque` will be `foo=bar;baz=qux` and ParseQuery admits both `&` and `;` separators.

Reply:
You are right it does https://golang.org/src/net/url/url.go?s=24541:24579#L880 I must have misread it before.